frontend/userpatch.lua

```diff
@@ -0,0 +1,121 @@
+-- return true if a patch was executed
+local function runUserPatchTasks(dir, priority)
+    if lfs.attributes(dir, "mode") ~= "directory" then
+        return
+    end
+
+    local patches = {}
+    for entry in lfs.dir(dir) do
+        local mode = lfs.attributes(dir .. "/" .. entry, "mode")
+        if entry and mode == "file" and entry:match("^" .. priority .. "%d*%-") then
+            table.insert(patches, entry)
+        end
+    end
+
+    if #patches == 0 then
+        return -- nothing to do
+    end
+
+    local function addLeadingZeroes(d)
+        local dec, n = string.match(d, "(%.?)0*(.+)")
+        return #dec > 0 and ("%.12f"):format(d) or ("%s%03d%s"):format(dec, #n, n)
+    end
+    local function sorting(a, b)
+        return tostring(a):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#b)
+            < tostring(b):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#a)
+    end
+
+    table.sort(patches, sorting)
+
+    for i, entry in ipairs(patches) do
+        local fullpath = dir .. "/" .. entry
+        if lfs.attributes(fullpath, "mode") == "file" then
+            if fullpath:match("%.lua$") then -- execute patch-files first
+                logger.info("User patch apply:", fullpath)
+                local ok, err = pcall(dofile, fullpath)
+                if not ok then
+                    logger.warn("User patch error applying patch:", fullpath, err)
+                    -- Only show InfoMessage, when UIManager is working
+                    if priority >= userpatch.late and priority < userpatch.before_exit then
+                        -- Only developers (advanced users) will use this mechanism.
+                        -- A warning on a patch failure after an OTA update will simplify troubleshooting.
+                        local UIManager = require("ui/uimanager")
+                        local InfoMessage = require("ui/widget/infomessage")
+                        UIManager:show(InfoMessage:new{text = "Error applying patch:\n" .. fullpath}) -- no translate
+                    end
+                end
+            end
+        end
+    end
+    return true
+end
+
+--- This function applies lua patches from `/koreader/userpatches`
+---- @string priority ... one of "early\_once", "early", "late", "before\_exit", "on\_exit"
+function userpatch.applyPatches(priority)
+    local patch_dir = data_dir .. "/userpatches"
+
+    if priority == userpatch.early then
+        -- Move an existing `koreader/patch.lua` to `koreader/userpatches/1-patch.lua` (-> will be excuted in `early`)
+        if lfs.attributes(data_dir .. "/patch.lua", "mode") == "file" then
+            if lfs.attributes(patch_dir, "mode") == nil then
+                if not lfs.mkdir(patch_dir, "mode") then
+                    logger.err("User patch error creating directory", patch_dir)
+                end
+            end
+            os.rename(data_dir .. "/patch.lua", patch_dir .. "/" .. userpatch.early .. "-patch.lua")
+        end
+    end
```

Comment:
Shouldn't this whole migration code be moved to OneTimeMigration?

Reply:
I decided not to do that, as onetime migration is executed later than this pice of code.

But if you think its no problem to move this to otm I have no problem with it (and I can reintroduce "rm afterupdate.markwr " there).

Reply:
At worst stuff is slightly broken on first boot?

Yeah, I feel like this just makes it seem more complicated than it ought to for something that'll essentially be dead code 99.99999% of the time ;).

Reply:
Then I will put that into onetime_migration. As we have room then, shall I add the `os.remove(afterupdete.marker)` again?